src/Week7/JoinList.hs

```diff
@@ -24,6 +24,18 @@ tag (Append m _ _) = m
 (+++) jl Empty = jl
 (+++) jl1 jl2  = Append (tag jl1 <> tag jl2) jl1 jl2
 
+---------- ex 0 ----------
+instance Monoid m => Semigroup (JoinList m a) where
+    (<>) = (+++)
+-- does this qualify as associative?
+-- (a <> b) + c = a <> (b <> c)
+-- only if we consider this to be true
+-- Append m (Append n a b) c = Append m a (Append p b c)
+-- which they are if we understand them as list waiting to be appended
+
+instance Monoid m => Monoid (JoinList m a) where
+    mempty = Empty
```

Comment:
@ChShersh  I'm not sure this is what you had in mind

Reply:
@JolandaNava I had in mind exactly the same 👍 The problem with things like `Append m (Append n a b) c = Append m a (Append p b c)` is typical problem for such `Semigroup` instance. The general agreement that it's okay while your interface doesn't see difference between `Append m (Append n a b) c` and `Append m a (Append p b c)`